pkg/gadgets/trace/tcpdrop/tracer/tracer.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2019-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tracer
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	"golang.org/x/sys/unix"
+
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/tcpdrop/tracer/tcpbits"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/tcpdrop/types"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target $TARGET -cc clang -no-global-types -type event tcpdrop ./bpf/tcpdrop.bpf.c -- -I./bpf/ -I../../../../${TARGET}
+
+type Config struct{}
+
+type Tracer struct {
+	config        *Config
+	enricher      gadgets.DataEnricherByMntNs
+	eventCallback func(*types.Event)
+
+	objs tcpdropObjects
+
+	kfreeSkbLink link.Link
+
+	reader *perf.Reader
+}
+
+func NewTracer(config *Config, enricher gadgets.DataEnricherByMntNs,
+	eventCallback func(*types.Event),
+) (*Tracer, error) {
+	t := &Tracer{
+		config:        config,
+		enricher:      enricher,
+		eventCallback: eventCallback,
+	}
+
+	if err := t.install(); err != nil {
+		t.close()
+		return nil, err
+	}
+
+	go t.run()
+
+	return t, nil
+}
+
+func (t *Tracer) close() {
+	t.kfreeSkbLink = gadgets.CloseLink(t.kfreeSkbLink)
+
+	if t.reader != nil {
+		t.reader.Close()
+	}
+
+	t.objs.Close()
+}
+
+func (t *Tracer) install() error {
+	spec, err := loadTcpdrop()
+	if err != nil {
+		return fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	gadgets.FixBpfKtimeGetBootNs(spec.Programs)
+
+	opts := ebpf.CollectionOptions{}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		var errVerifier *ebpf.VerifierError
+		if errors.As(err, &errVerifier) {
+			fmt.Printf("Verifier error: %+v\n",
+				errVerifier)
+		}
```

Comment:
mmm, I know this is important to understand when there is a failure, however we're not doing this on other gadgets. Do you think we should make this on all gadgets?

Reply:
I think we should have a way to enable this on all gadgets. But it's unrelated to this PR so I will remove it to keep consistent.